vars/FreeBSD.yml

```diff
@@ -0,0 +1,4 @@
+---
+nginx_conf_dir: /usr/local/etc
+nginx_user: www
+nginx_html_file_location: /usr/local/www/nginx-dist
```

Comment:
In the following statement your intention was probably to comment on the variable **html_file_location**. Let me quote it:

> "The original idea behind the nginx_html_file_location variable was to allow having multiple location blocks all pointing to different HTML file locations if the user had multiple HTML file locations."

Atm the variable html_file_location is used in nginx_html_demo_template and nginx_http_template only. An option would be to use a dictionary

    nginx_html_file_location:
      demo: /usr/local/www/nginx-demo
      basic: /usr/local/www/nginx-dist

and fix the defaults accordingly
        html_file_location: "{{ nginx_html_file_location.basic }}"

This structure would fit your intention better, I think. But before you make a decision it would be good to take into the account all aspects of the "OS specific variables" proposed by other patch (tasks/vars.yml).

**Precedence**
"OS specific variables" are [include_vars](https://docs.ansible.com/ansible/latest/modules/include_vars_module.html) read by tasks/vars.yml. include_vars precedence  is rather high. [See the table](https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#variable-precedence-where-should-i-put-a-variable). Only couple of options left to override these variables.

**Use case**
Atm most of the variables in **nginx_http_template** and in **nginx_html_demo_template** are hard-coded. For a user to change these variables means to copy the whole templates into higher precedence and customize it there. An option would be to introduce default variables from files in vars/default (see below).  The customized files then might be placed in the vars directory.

(tasks/vars.yml)

    - name: (Setup: All OSs) Set default variables
      include_vars: "{{ item }}"
      with_first_found:
        - files:
            - "{{ ansible_distribution }}-{{ ansible_distribution_release }}.yml"
            - "{{ ansible_distribution }}.yml"
            - "{{ ansible_os_family }}.yml"
            - "default.yml"
          paths: "{{ role_path }}/vars/default"

    - name: (Setup: All OSs) Set custom variables
      include_vars: "{{ item }}"
      with_first_found:
        - files:
            - "{{ ansible_distribution }}-{{ ansible_distribution_release }}.yml"
            - "{{ ansible_distribution }}.yml"
            - "{{ ansible_os_family }}.yml"
            - "default.yml"
          paths: "{{ role_path }}/vars"

I think this way it's possible to be flexible and in the same time 1) not to break current features and 2) not to make it harder to use. Ansible si flexible and there are many other possibilities. Take your time and try to find out how you would like to see your customers to customize the role. It's kind of crossroad and It will be harder to change it later. If you have any questions don't hesitate.

Meanwhile I'll prepare other changes you recommended.

Reply:
Thanks for the comprehensive explanation! However, I think you misunderstood what I meant. Let me try to explain using a mock playbook 😛
```yaml
---
- hosts: localhost
  become: true
  remote_user: root
  roles:
    - ansible-role-nginx
  vars:
    nginx_http_template_enable: true
    nginx_http_template:
      app:
        template_file: http/default.conf.j2
        conf_file_name: default.conf
        conf_file_location: /etc/nginx/conf.d/
        port: 80
        server_name: localhost
        error_page: /usr/share/nginx/html
        reverse_proxy:
          locations:
            frontend:
              location: /
              proxy_pass: http://frontend_servers/
            backend:
              location: /backend
              proxy_pass: http://backend_servers/
        upstreams:
          upstream_1:
            name: frontend_servers
            lb_method: least_conn
            zone_name: frontend
            zone_size: 64k
            sticky_cookie: false
            servers:
              frontend_server_1:
                address: localhost
                port: 8081
                weight: 1
                health_check: max_fails=3 fail_timeout=5s
          upstream_2:
            name: backend_servers
            lb_method: least_conn
            zone_name: backend
            zone_size: 64k
            sticky_cookie: false
            servers:
              backend_server_1:
                address: localhost
                port: 8082
                weight: 1
                health_check: max_fails=3 fail_timeout=5s
      frontend:
        template_file: http/default.conf.j2
        conf_file_name: frontend_default.conf
        conf_file_location: /etc/nginx/conf.d/
        port: 8081
        server_name: localhost
        error_page: /usr/share/nginx/html
        autoindex: false
        web_server:
          locations:
            default:
              location: /
              html_file_location: /usr/share/nginx/frontend_html
              html_file_name: frontend_index.html
              autoindex: false
          http_demo_conf: true
      backend:
        template_file: http/default.conf.j2
        conf_file_name: backend_default.conf
        conf_file_location: /etc/nginx/conf.d/
        port: 8082
        server_name: localhost
        error_page: /usr/share/nginx/html
        autoindex: false
        web_server:
          locations:
            default:
              location: /
              html_file_location: /usr/share/nginx/backend_html
              html_file_name: backend_index.html
              autoindex: false
          http_demo_conf: true
    nginx_html_demo_template_enable: true
    nginx_html_demo_template:
      frontend:
        template_file: www/index.html.j2
        html_file_name: frontend_index.html
        html_file_location: /usr/share/nginx/frontend_html
        web_server_name: Frontend
      backend:
        template_file: www/index.html.j2
        html_file_name: backend_index.html
        html_file_location: /usr/share/nginx/backend_html
        web_server_name: Backend
```
In this case there's two different folder locations, one for frontend HTML sites and one for backend HTML sites. I want to make sure that it's still possible to easily use different folder locations. FWIW, and to clarify the matter a little bit further, the `html_file_location` within the `nginx_http_template` block refers to the location of an existing HTML file, whereas the `html_file_location` in the `nginx_html_demo_template` block refers to the location where a demo NGINX HTML file will be create.

Reply:
Thank you for the playbook. You haven't answered the question: "How would you like to customize the role?" For now would you like to proceed with the "OS specific variables" at the playbook's level?

Reply:
> Thank you for the playbook. You haven't answered the question: "How would you like to customize the role?" For now would you like to proceed with the "OS specific variables" at the playbook's level?

 I think OS specific variables make sense for some variables such as `nginx_conf_dir` or `nginx_user` that are usually very specific for any given OS. I do not think they make that much sense for values such as `html_file_location` (or at least, not in the current form). Setting a default value for a value such as `nginx_http_template.x.web_server.locations.x.html_file_location` would only work if it maintains the current functionality of having a potentially unlimited number of different `html_file_locations`.

> Use case
Atm most of the variables in nginx_http_template and in nginx_html_demo_template are hard-coded. For a user to change these variables means to copy the whole templates into higher precedence and customize it there. An option would be to introduce default variables from files in vars/default (see below). The customized files then might be placed in the vars directory.

I'm not sure I follow. Do you mean copying the template? I am not entirely sure why a user would want to do that given how relatively simple it would be to change the original template once the role has been downloaded.

If you mean that users right now have to create a playbook that defines each variable within the template or else the playbook will fail then yeah it is definitely in my mind that right now users need to copy the whole template into a playbook and configure each variable or else the playbook will fail to run. Following from what I said above, I'm not against implementing specific default variables at a high level, as long as it maintains current functionality (eg a default like `html_file_location: "{{ nginx_html_file_location.basic }}"` would not work since there's only ever going to be one value for `html_file_location: "{{ nginx_html_file_location.basic }}"`). Otherwise the other solution I had in mind to fix this issue was to use the `default` keyword to set sensible defaults - this wouldn't break any of the current functionality.

Hopefully that makes sense? Let me know what you think 😃